### PR TITLE
feat(g-code-testing): HTTP G-Code GitHub Action (RET-174)

### PR DESCRIPTION
**Pull request by DerekMaggio**
# Overview

Add HTTP tests to Github action
Refactor framework to use pytest

# Changelog
- Add AWS python package, `boto3`, to `g-code-testing` project
- Revert passing classes to G-Code Engine and instead just pass parameters required for execution
- Remove GCodeTestData class and configurations file and any associated references/tests
- Remove cli and any associated references/tests
- Update HTTP and protocol G-Code confirmation tests to work with pytest
- Add pytest marks to tests to classify them
  - Add `slow` custom mark
  - Add `g_code_confirm` custom mark
- Update Makefile `test` targets
  - Classifications of tests
    1. Framework Unit Tests - tests to ensure that framework has not regressed - under normal `make test` call 
    2. Fast G-Code Confirm Tests - G-Code Confirmation tests that run quickly (sub 1 minute) - under `test-g-code-fast`
    3. Slow G-Code Confirm Tests - G-Code Confirmation tests that run quickly (above 1)  - separated out into specific targets per test to allow for parallelization in Github Action. For example, `make test-g-code-swift-smoke`
- Consolidate `g-code-testing` project into 2 actions
  - `G-Code Testing Lint & Test` - Run linting and `Framwork Unit Tests`
  - `G-Code-Confirm` - Run `Fast G-Code Confirm Tests` in single job and each slow test in it's own job (Each job runs in parallel)

# Review requests

Take a look at the README and note the Control Tests section. Does it make sense to skip the tests that I noted as skipped?

# Risk assessment

Could break G-Code Confirmation tests. Rerunning tests in CI should take care of this though

### CODE REVIEW

File: .github/workflows/g-code-http-tests.yaml [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  pull_request:
+    paths:
+      - 'api/**'
+      - 'g-code-testing/**'
+      - 'Makefile'
+    branches:
+      - '*'
+  push:
+    paths:
+      - 'api/**'
```
Comment (sfoster1) [inline .github/workflows/g-code-http-tests.yaml]: You probably want to trigger it on this workflow file too
```suggestion
      - 'api/**'
      - '.github/workflows/g-code-http-tests.yaml'
```

File: .github/workflows/g-code-http-tests.yaml [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        with:
+          python-version: 3.7
+
+      - name: Setup G-Code Testing Project
+        uses: ./.github/actions/python/setup
+        with:
+          project: g-code-testing
+
+      - name: Run HTTP call - Move Left Pipette
+        if: always()
```
Comment (sfoster1) [inline .github/workflows/g-code-http-tests.yaml]:
```suggestion
```

Do you need this?

**Review by sfoster1 — CHANGES_REQUESTED** [review]
This is a lot of tests, and they're all separate makefile targets and each makefile target is a separate line in the github action. That's a lot to keep up to date and maintain; in addition to whatever's necessary to make the test run, it guarantees touching two other files.

What if instead we used a test system with test autodiscovery - pytest, for instance, since this is in python - and used its output reporting? one command, and the only thing you need to touch to change or add a test is the test code itself.

**Comment by DerekMaggio** [comment]
> This is a lot of tests, and they're all separate makefile targets and each makefile target is a separate line in the github action. That's a lot to keep up to date and maintain; in addition to whatever's necessary to make the test run, it guarantees touching two other files.
> 
> What if instead we used a test system with test autodiscovery - pytest, for instance, since this is in python - and used its output reporting? one command, and the only thing you need to touch to change or add a test is the test code itself.

@sfoster1, the hard part is comparing against S3.
Is it okay if I pull in the S3 python package and use that?

We will also then have to think about how to get the credentials into the Python program. I can probably do it with env vars to the program.

File: g-code-testing/tests/g_code_comparision/test_quick_protocol.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        "2-single-channel-v2.txt",
+        "protocols/2_single_channel_v2.py",
+        two_single_channel_setup(),
+    ],
+]
+
+
+@pytest.mark.g_code_confirm
+@pytest.mark.parametrize("master_file_name,path,settings", TEST_DATA)
+def test_quick_protocols(master_file_name, path, settings):
```
Comment (amitlissack) [inline g-code-testing/tests/g_code_comparision/test_quick_protocol.py]: Would be nice to see type annotations on all args and returns.

Comment (DerekMaggio) [reply g-code-testing/tests/g_code_comparision/test_quick_protocol.py]: Updated

File: .github/workflows/g-code-confirm-tests.yaml [hunk 4]
```diff
@@ -0,0 +1,112 @@
+name: "G-Code-Confirm"
```
Comment (sfoster1) [inline .github/workflows/g-code-confirm-tests.yaml:1]: I think you can use https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix to make this a bit more concise. It's just text substitution, so you could make your matrix a list of tests to run and put it in the commands to make and the names of the jobs

**Review by sfoster1 — APPROVED** [review]
This looks great! I think you can make the yaml a bit better using job matrices, but that's about it. Really like the use of pytest markers.

**Pull request merged**